feat(blog): "What are Durable Functions? A visual JavaScript primer" article (#916)
inngest/website: pages/blog/_posts/durable-functions-a-visual-javascript-primer.mdx

---
focus: true
heading: "What are Durable Functions? A visual JavaScript primer"
image: /assets/blog/durable-functions-a-visual-javascript-primer/featured-image.png
showSubtitle: false

> subtitle: "Lydia Hallie's powerfull animated illustrations introduce us to the inner workings of Durable Functions."

Author: ```suggestion
subtitle: "Lydia Hallie's powerful animated illustrations cover the inner workings of Durable Functions."
```

- **Scalability Concerns**: As the number of background jobs grows, it becomes increasingly challenging to manage their execution, monitor them, and handle errors.

</Callout>

---

## Durable Functions

Now, let's reimagine the same e-commerce platform using Durable Functions. But first, what *are* Durable Functions?

The term "durable" in Durable Functions refers to the workflow's ability to maintain state and logic through **failures**, **retries**, and **pauses**. A Durable Function **guarantees a start-to-finish execution**; once it's started, it will eventually complete its execution regardless of any unexpected issues. 

![Durable Functions are designed with reliability at their core with automatic retries.](/assets/blog/durable-functions-a-visual-javascript-primer/Group_85_(1).png)

This is what makes Durable Functions special, because this wasn't the case with our traditional setup. 

---

In a Durable Function, each task in the order processing workflow is broken down into modular, independent **steps**. For example, the order processing workflow could include steps for validating payment, checking inventory, scheduling shipment, and updating the customer. 

> Inngest's SDK allows you to create steps through the [`step.run`](http://step.run) function. 

Reviewer: Let's not link this yet.
```suggestion
Inngest's SDK allows you to create steps through the `step.run` function. 
```

  { id: "Order Processing" },
  { event: "order.placed" },
  async ({ step, event }) => {
    const paymentConfirmation = await step.run("process-payment", () =>
      processPayment(event.data.orderDetails)
    );
    
    const inventoryStatus = await step.run("check-inventory", () =>
      checkInventory(event.data.orderDetails.items)
    );

     const shipmentDetails = await step.run("schedule-shipment", () =>
      scheduleShipment(event.data.orderDetails)
    );
    
    await step.run("send-notification", () =>
      sendNotification(
        event.data.customerId,
        "Order Received",
        shipmentDetails.status
      )
    );
  }
);
```

Each step is capable of handling failures and can automatically retry based on predefined policies, ensuring that all parts of the order processing sequence is executed correctly. 

Durable Functions can also include **waits** that pause the process for necessary delays (e.g., waiting for a third-party shipment confirmation) without consuming resources. This can be more efficient than traditional CRON jobs, which are less time flexible and can over-consume resources by continuously polling for status updates.

> Inngest's SDKS allows you to create waits through several methods, one of which is the [`step.waitForEvent`](https://www.inngest.com/docs/reference/functions/step-wait-for-event) function.  

Reviewer: ```suggestion
Inngest's SDKS allows you to create waits through several methods, one of which is the [`step.waitForEvent`](/docs/reference/functions/step-wait-for-event?ref=blog-what-are-durable-functions) function.  
```

      event: "shipment/create",
      timeout: "7d",
      if: "event.data.id == async.data.orderId",
    });

    if (shipment) {
      await step.run("send-notification", () =>
        sendNotification(event.data.customerId, "Order Shipped")
      );
    }
  }
);
```

In this workflow, after scheduling the shipment, the workflow enters a wait state up to 7 days using the `step.waitForEvent` function. The `step.waitForEvent` function enables our workflow to wait for a new `"shipment.created”` matching the same `orderId`. This wait pauses the workflow until the logistics provider confirms the shipment has been dispatched. Unlike regular background jobs, this wait doesn't use resources while it waits, making it better for handling time-sensitive tasks in web apps.

<Callout>

💡 To summarize, our Durable Function approach enhances the efficiency and reliability of long-running tasks compared to the traditional approach:

- **Robust State Management**: Each step in a Durable Function is capable of maintaining its state between executions. This prevents the need to restart tasks from the beginning if they fail, minimizing the risk of duplication or inconsistent results.

> - **Advanced Error Handling**: Durable Functions are equipped with built-in mechanisms for [error detection and automatic retries](https://www.inngest.com/docs/guides/error-handling), which can be tuned to address specific recovery scenarios. This not only improves the reliability of task execution, but also reduces the manual oversight that was required with the traditional approach.

Reviewer: ```suggestion
- **Advanced Error Handling**: Durable Functions are equipped with built-in mechanisms for [error detection and automatic retries](/docs/guides/error-handling?ref=blog-what-are-durable-functions), which can be tuned to address specific recovery scenarios. This not only improves the reliability of task execution, but also reduces the manual oversight that was required with the traditional approach.
```

- **Efficient Long-Running Process Management**: Each task within a Durable Function is treated as a modular step that can independently handle long durations of activity. This modularity allows for more scalable and manageable processes, overcoming the limitations of traditional background jobs that are typically optimized for shorter tasks.
- **Enhanced Scalability**: As the complexity and number of tasks grow within an application, Durable Functions provide a scalable framework that simplifies monitoring and managing these tasks. The independent steps ensure that the performance and stability of one does not affect others, which is a common challenge with traditional background jobs.

</Callout>

---

> ## Durable Functions

Author: ```suggestion
## Durable Functions: under the hood
```

> ## Durable Functions: under the hood

Author: ```suggestion
### Under the hood
```

> Durable Functions provide these features automatically, but how do they work behind the scenes? Let's explore what happens when we run our e-commerce code snippet step-by-step using Inngest's Durable Execution Engine and SDK.

Reviewer: SEO backlink
```suggestion
Durable Functions provide these features automatically, but how do they work behind the scenes? Let's explore what happens when we run our e-commerce code snippet step-by-step using Inngest's [Durable Execution Engine](/docs/learn/how-functions-are-executed?ref=blog-what-are-durable-functions) and SDK.
```

Before we start, we first need to understand that:

- The Inngest Platform hosts the Durable Execution Engine. The engine is responsible for managing the queue of pending function runs, and maintaining the state of ongoing or pending Function steps.
- The Durable Execution Engine interacts with the Inngest SDK through HTTP requests. Through this communication, the engine provides states and collects updated states from the SDK.

> - The functions need to be [served through HTTP endpoints](https://www.inngest.com/docs/learn/serving-inngest-functions). In this step-by-step example, let's assume that we've successfully done this.

Reviewer: ```suggestion
- The functions need to be [served through HTTP endpoints](/docs/learn/serving-inngest-functions?ref=blog-what-are-durable-functions). In this step-by-step example, let's assume that we've successfully done this.
```
